Add template-level parameters and `file_name` per template config

lquerel opened the pull request:
This PR adds two new optional fields in the `templates` section of a Weaver configuration file:
- `params`: Defines a map of key-value pairs which will be merged with the ones defined at the top level of the `weaver.yaml` file. CLI parameters can also be used to override these parameters.
- `file_name`: Defines the relative file path of the output generated from the template. This field is a Jinja expression which will be evaluated with the same context used in the template itself. The parameters mentioned above are also accessible.

An example of usage:
```yaml
params:
  exclude_deprecated: true
  
templates:
  - pattern: attribute.py
    params:
      exclude_stability: ["stable"]
    filter: "semconv_grouped_attributes($params)"
    application_mode: each
    file_name: "incubating/{{ctx.root_namespace | snake_case}}.py"
  - pattern: attribute.py
    params:
      exclude_stability: ["experimental"]
    filter: "semconv_grouped_attributes($params)"
    application_mode: each
    file_name: "{{ctx.root_namespace | snake_case}}.py"
```

Note: `params` are now merged in an additive way. For example, if a parameter is defined in the `params` section of a template, then the final `params` section for this template will include both the `params` inherited from the file-level configuration and the `params` defined in the template configuration. If the override is a parameter set to null, then this parameter is removed from the final `params` section. This modification should not have a significant impact on the existing codegen, as I don't believe any codegen was relying on this specific behavior.

Closes [open-telemetry/weaver#288](https://github.com/open-telemetry/weaver/issues/288).

jsuereth commented on crates/weaver_semconv_gen/Cargo.toml:
```diff
@@ -18,7 +18,6 @@ weaver_resolved_schema = { path = "../weaver_resolved_schema" }
 weaver_semconv = { path = "../weaver_semconv" }
 nom = "7.1.3"
 
-itertools.workspace = true
```
Think I missed this when removing the `writeln!` code, good catch.

jsuereth commented on crates/weaver_forge/src/lib.rs:
```diff
@@ -192,13 +198,33 @@ impl TemplateEngine {
+                // Note: The result of the insert method is ignored because we don't care about
+                // the previous value of the param.
+                for (name, value) in params.params.iter() {
+                    _ = template_params.insert(name.clone(), value.clone());
```
Nit: can you add the comment about ignoring the previous value here?

lquerel replied:
Done

jsuereth reviewed (commented):
Overall looks good BUT I find the weaver config confusing.

I think we should deprecated `pattern` for `template`.

Specifically:

```
templates:
  - pattern: test.md
    filter: .
    application_mode: single
    params:
      template_name: test1
      test_1: true
    file_name: "{{params.template_name | snake_case}}.json"
```

Looks confusing, since `test.md` is the *template* you're using, not a pattern at all.

Instead, let's do:

```
templates:
  - template: test.md
    filter: .
    application_mode: single
    params:
      template_name: test1
      test_1: true
    file_name: "{{params.template_name | snake_case}}.json"
```

I think we can do this in a non-breaking way by deprecating the "pattern" attribute and adding "template" where we pull from pattern if it exists, otherwise use template.

lquerel commented:
> I think we can do this in a non-breaking way by deprecating the "pattern" attribute and adding "template" where we pull from pattern if it exists, otherwise use template.

Can be done with a serde alias. I will also update the docs.

lquerel commented:
@jsuereth I renamed `templates.pattern` to `templates.template` and added `pattern` as a serde alias of `template` for backward-compatibility. I also updated documentations and examples.
